### PR TITLE
Add hook for loading certificate and key data from plugin 

### PR DESCRIPTION
**Pull request by shinrich**
Note: I'm marking this for consideration for 9.0.  I am ok with it not going into the first version.  My organization needs this functionality, but we can patch it in for a bit.  Would much rather get some version of 9.0 released sooner.

This branch adds functionality for plugins to load certificate and key information on configuration load and reload.  The new plugin hook triggers for both client certificates and server certificates,

This branch adds a hook (TS_LIFESTYLE_SSL_SECRET_HOOK) and a pair of TS API's, TSSslSecretSet/Get.  The hook gets triggered before the SSL_CTX's are created.  The continuation is called with the name of the certificate or key as the data argument.  The plugin then can load the certificate/key information from some other source (e.g. a key server or just someplace else on disk as the ssl_secret_load_test.cc does in this branch).  The plugin calls TSSslSecretSet to update the secret table.  After the hooks are called, the core looks for the data in the secret_map.  If it is not present, the core loads the data from disk and puts it in the map.  The SSL_CTX creation logic was updated to operate on a mem_bio instead of using the file oriented forms of the cert and key loading calls.

For the client certs specified by conf_remap.so, the secrets are loaded as needed likely after the config load/reload phase.  Same plugin flow applies.

The branch adds a pre-step for the config load so the secrets are loaded before the regular SSL_CTX creation is performed.

The branch adds a secret_map which is effectively filled with the pem encoded certificates and keys.  This is an increase in memory foot print.  The core logic could clear the secret entry after creating the SSL_CTX, but of the same certificate/key is used for multiple SSL_CTX, this will result in extra calls to the plugin or disk.  For config load/reload we could just clear the secret map after the config load stage is done.  For conf_remap updates, the getCTX table should have the same SSL_CTX if the certificate/key is used multiple times, so freeing the entries after the individual load should be fine.

I will be sending a mail out on the dev list highlighting the API change and pointing to this PR.

After we get the client and server side plugin-based certificate loading working, the next step is allowing a plugin to push in changes without a full config reload.

### CODE REVIEW

**Comment by randall** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
I need to go back and make this all work with Quic which is why the autests are failing.

**Comment by shinrich** [comment]
Adding server certificate support wasn't that hard with the other preparation work in place, so I added it in.  Now off to make this work with quic.

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
Pushed another commit to track version of the secrets.  When testing with our secret server, my plugin needed to share the version between the certificate and the key secrets to make sure they were the same version.

File: iocore/net/SSLSecret.cc [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  }
+  return &iter->second;
+}
+
+bool
+SSLSecret::getSecret(const char *name, int &version, const char **data, int *data_len) const
+{
+  const SSLSecretData *data_item = this->getSecret(name);
+  if (data_item) {
+    Debug("ssl_secret", "Get secret=%10.s...  %s(%" PRId64 ") version %d", name, data_item->data.data(), data_item->data.length(),
```
Comment (maskit) [inline iocore/net/SSLSecret.cc:110]: I got this error. `%zu` is probably the appropriate format specifier.
```
  CXX      SSLSecret.o
SSLSecret.cc:110:106: error: format specifies type 'long long' but the argument has type 'std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::size_type' (aka 'unsigned long') [-Werror,-Wformat]
    Debug("ssl_secret", "Get secret=%10.s...  %s(%" PRId64 ") version %d", name, data_item->data.data(), data_item->data.length(),
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
../../include/tscore/Diags.h:329:39: note: expanded from macro 'Debug'
      diags->log(tag, DL_Debug, &loc, __VA_ARGS__);    \
                                      ^~~~~~~~~~~
1 error generated.
```

File: doc/developer-guide/api/functions/TSSslSecret.en.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+.. code-block:: cpp
+
+    #include <ts/ts.h>
+
+.. function:: TSReturnCode TSSslSecretGet(const char * secret_name, const char **secret_data_return, int *secret_data_len)
+
+Description
+===========
+
+:func:`TSSslSecretGet` fetchs the named secret from the current secret map. TS_ERROR is returned if there is no entry for the secret.
```
Comment (randall) [inline doc/developer-guide/api/functions/TSSslSecret.en.rst:59]: fetches

File: doc/developer-guide/api/functions/TSLifecycleHookAdd.en.rst [hunk 3]
```diff
@@ -106,6 +106,14 @@ Types
       Invoked with the event :c:data:`TS_EVENT_LIFECYCLE_TASK_THREADS_READY` and ``NULL``
       data.
 
+   .. cpp:enumerator:: TS_LIFECYCLE_SSL_SECRET_HOOK
+
+      Called before the data for the certificate or key is loaded.  The data argument to the callback is a pointer to a :type:`TSSecretID` which
+      contains a pointer to the name of the certificate or key and the relavent version if applicable.
```
Comment (randall) [inline doc/developer-guide/api/functions/TSLifecycleHookAdd.en.rst:112]: relevant

File: tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("CN = group.com", "Should select a group SAN");
+
+# Should receive a RSA cert
+tr = Test.AddTestRun("rsa.com only in rsa cert")
+tr.Processes.Default.Command = "echo foo | openssl s_client -servername rsa.com -connect 127.0.0.1:{0}".format(ts.Variables.ssl_port)
+tr.ReturnCode = 0
+tr.StillRunningAfter = server
+tr.StillRunningAfter = ts
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("Peer signature type: RSA-PSS", "Should select RSA cert")
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("CN = group.com", "Should select a group SAN");
```
Comment (randall) [inline tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py:123]: trailing semicolon

File: tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+# Should receive a EC cert
+tr = Test.AddTestRun("ec.com only in ec cert")
+tr.Processes.Default.Command = "echo foo | openssl s_client -servername ec.com -connect 127.0.0.1:{0}".format(ts.Variables.ssl_port)
+tr.ReturnCode = 0
+tr.StillRunningAfter = server
+tr.StillRunningAfter = ts
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("Peer signature type: ECDSA", "Should select EC cert")
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("CN = group.com", "Should select a group SAN");
```
Comment (randall) [inline tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py:132]: trailing semicolon

File: tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+# Should receive a RSA cert
+tr = Test.AddTestRun("Only offer RSA ciphers, should receive RSA cert")
+tr.Processes.Default.Command = "echo foo | openssl s_client -servername one.com -sigalgs 'RSA-PSS+SHA256' -connect 127.0.0.1:{0}".format(ts.Variables.ssl_port)
+tr.ReturnCode = 0
+tr.StillRunningAfter = server
+tr.StillRunningAfter = ts
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("Peer signature type: RSA-PSS", "Should select RSA cert")
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("CN = group.com", "Should select a group SAN");
```
Comment (randall) [inline tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py:114]: trailing semicolon

File: tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("Peer signature type: RSA-PSS", "Should select RSA cert")
+
+# Should receive a EC cert
+tr = Test.AddTestRun("Default for one.com should return EC cert")
+tr.Processes.Default.Command = "echo foo | openssl s_client -servername one.com -connect 127.0.0.1:{0}".format(ts.Variables.ssl_port)
+tr.ReturnCode = 0
+tr.StillRunningAfter = server
+tr.StillRunningAfter = ts
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("Peer signature type: ECDSA", "Should select EC cert")
+tr.Processes.Default.Streams.All += Testers.ContainsExpression("CN = group.com", "Should select a group SAN");
```
Comment (randall) [inline tests/gold_tests/tls/tls_check_dual_cert_selection_plugin.test.py:105]: trailing semicolon

File: tests/gold_tests/tls/tls_client_cert_override_plugin.test.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+# Need to copy over the environment so traffic_ctl knows where to find the unix domain socket
+trupdate.Processes.Default.Env = ts.Env
+trupdate.Processes.Default.ReturnCode = 0
+
+
+# Parking this as a ready tester on a meaningless process
+# Stall the test runs until the sni reload has completed
+# At that point the new sni settings are ready to go
+def sni_reload_done(tsenv):
+  def done_reload(process, hasRunFor, **kw):
```
Comment (randall) [inline tests/gold_tests/tls/tls_client_cert_override_plugin.test.py:156]: Indentation here is not multiple of four

File: tests/gold_tests/tls/tls_client_cert_plugin.test.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+tr2.StillRunningAfter = server2
+tr2.Processes.Default.Command = 'echo Updated configs'
+# Need to copy over the environment so traffic_ctl knows where to find the unix domain socket
+tr2.Processes.Default.Env = ts.Env
+tr2.Processes.Default.ReturnCode = 0
+
+# Parking this as a ready tester on a meaningless process
+# Stall the test runs until the sni reload has completed
+# At that point the new sni settings are ready to go
+def sni_reload_done(tsenv):
```
Comment (randall) [inline tests/gold_tests/tls/tls_client_cert_plugin.test.py:184]: indentation here is not multiple of four

**Comment by shinrich** [comment]
Pushed a new version that includes all the update (hot load) as well as the initial load.  Basically all the features I described in the Spring 2020 summit. https://cwiki.apache.org/confluence/display/TS/Presentations+-+2020?preview=/158863592/158864749/ATS%20Summit%20Cert%20Loading.pdf

**Comment by shinrich** [comment]
Yes, that is a concern.  The keys are already being kept in memory anyway as part of the SSL_CTX, but the secret map keeps it there twice.  Seems like we should be able to flush the data out sooner, but will probably need some additional meta data to track changes, etc.  I'll take another pass on that.

**Comment by maskit** [comment]
Yeah, SSL_CTX keeps keys and it's inevitable. However, seems like they are carefully managed. I didn't look into the detail but found these functions. Using these functions may mitigate the risks.
https://www.openssl.org/docs/man1.1.1/man3/OPENSSL_secure_free.html

**Comment by randall** [comment]
@shinrich When you get a chance, can you rebase this PR?

**Comment by shinrich** [comment]
@randall finally got this PR rebased.

**Comment by shinrich** [comment]
[approve ci clang-format]

**Comment by shinrich** [comment]
[ approve ci clang-format ]

**Comment by shinrich** [comment]
[approve ci clang-format]

**Comment by shinrich** [comment]
[approve ci clang-format]

**Review by maskit — COMMENTED** [review]
Besides the security concern, it looks good.

Would it be a big change if we stored actual key data in secure heap (by using OPENSSL_secure_malloc) and only put the pointers to secret_map? I think it's relatively easier than managing TTL.

I could approve this, but I think this needs to get a couple of more eyes.

**Comment by shinrich** [comment]
OPENSSL_secure_malloc looks interesting.  Would be easy enough to use. I'll poke through the implementation to understand what it does on our architectures.

**Comment by shinrich** [comment]
[approve ci ubuntu]

**Comment by shinrich** [comment]
[approve ci clang-format]

**Comment by shinrich** [comment]
Checked out the OPENSSL_secure_malloc implementation.  OPENSSL_secure_malloc creates a separate heap to for OPENSSL_secure_malloc to allocate from.  The logic is that basic heap overflows from regular memory allocations won't override into your sensitive data.

The secure heap is fixed size and all allocs and frees to that heap are serialized by a mutex.  So we would definitely want to do another performance evaluation before committing that change. 

I would prefer to land this change which we have been running in production for at least a year.  Then spend some effort to reduce the lifetime of the secrets in memory.  I think that will have a bigger security benefit than the secure heap.  That will involve adjusting our reloading/polling logic and our lazy loading logic (for client certificates).

**Review by maskit — DISMISSED** [review]
Maybe I'm paranoid, so If we really want to have this on 9.1.0 then go ahead. If not, I don't see the point to land this soon without taking any effort to make this secure - I don't have a strong opinion on how to do it. Please note that we cannot revert this change easily once we ship it as a release because this introduces new TS APIs.

File: tests/tools/plugins/ssl_secret_load_test.cc [hunk 10]
```diff
@@ -0,0 +1,184 @@
+/** @file
+
+  SSL Preaccept test plugin
+  Implements blind tunneling based on the client IP address
+  The client ip addresses are specified in the plugin's
```
Comment (zwoop) [inline tests/tools/plugins/ssl_secret_load_test.cc:5]: Is this comment copied from another plugin?

**Review by zwoop — APPROVED** [review]

**Comment by shinrich** [comment]
[approve ci]

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
Hmm trying again.  CI doesn't seem to be picking up my latest commit which fixes the tests

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
Trying again, the autest last time did not recognize the new commit "error: no such commit c83249ec762b7e038a2ba552cdf9ebf6d2926ac5"

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
@randall is going to take a look at it.

**Review by randall — APPROVED** [review]

**Pull request merged**